### ISSUE TITLE
BYO moment.js

### ISSUE THREAD

**Issue by lb-** [lb-/moment-helpers#15]
Posted this on the forum https://forums.meteor.com/t/advice-for-building-a-meteor-moment-js-helpers-package/1299

Essentially, this package should not require a specific version of meteor or package but allow the user to Bring their own Meteor package

**Comment by chicagogrooves** [comment]
For that matter, I wonder why I so frequently get this error as I develop:
```
Error: lbee:moment-helpers, no Moment package is installed
```
I see your `packge.js` pulls in moment, yet while rebooting I frequently get:

````
W20150404-03:14:04.995(-5)? (STDERR) Error: lbee:moment-helpers, no Moment package is installed
W20150404-03:14:04.995(-5)? (STDERR)     at Package (packages/lbee:moment-helpers/common.js:152:1)
W20150404-03:14:04.995(-5)? (STDERR)     at /Users/dradcliffe/src/opedia/app/.meteor/local/build/programs/server/packages/lbee_moment-helpers.js:175:4
W20150404-03:14:04.995(-5)? (STDERR)     at /Users/dradcliffe/src/opedia/app/.meteor/local/build/programs/server/packages/lbee_moment-helpers.js:184:3
W20150404-03:14:04.996(-5)? (STDERR)     at /Users/dradcliffe/src/opedia/app/.meteor/local/build/programs/server/boot.js:222:10
W20150404-03:14:04.996(-5)? (STDERR)     at Array.forEach (native)
W20150404-03:14:04.996(-5)? (STDERR)     at Function._.each._.forEach (/Users/dradcliffe/.meteor/packages/meteor-tool/.1.1.1.46baqh++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/server-lib/node_modules/underscore/underscore.js:79:11)
W20150404-03:14:04.996(-5)? (STDERR)     at /Users/dradcliffe/src/opedia/app/.meteor/local/build/programs/server/boot.js:117:5
```

Let me know if this merits its own issue.

**Comment by lb-** [comment]
@chicagogrooves FYI I have removed that checking and for now am resolving that BYO moment.js is not something I can implement with the current Meteor packages api, this should stop that error from appearing.

**Comment by chicagogrooves** [comment]
Hmm. If you say a little more about that, I could try and look into it. But thanks for the patch for now, that's helpful.

**Comment by lb-** [comment]
@chicagogrooves I have reopened, this and have a renewed passion to give this another go, I would love your thoughts.

Firstly, on the post https://forums.meteor.com/t/advice-for-building-a-meteor-moment-js-helpers-package/1299/9 serkandurusoy suggested I try to put this code in package.js

```
if (!moment) {
  api.use(moment@x.x.x, [client,server])
}
```

Secondly, I think the issue was that I was checking for moment when the package was loaded, but if I am using a weak dependency or even NO dependency, I should only check when the mo function is used, throwing an error then (as it would mean that moment does not exist AFTER all packages are loaded).

Still learning & would love your feedback on this approach, will try to test it out in the next little while.

**Comment by serkandurusoy** [comment]
The thing is, you actually do depend on moment so you need to make sure your package loads after moment packages and the only way to do that is to declare your moment dependency. But you also don't want to bring in moment yourself. So this causes a dilemma. I think there is no actual clean solution here yet. The weak dependency works only if you can live without the dependency and can make use of it if it is present.

My suggestion with the package js solves the issue when there is no moment package at all. But if there is, or there is one but loaded after yours, then there still will be a problem with version constraints.

Perhaps an intermediary solution would be to take a note on the readme and require the app developer to bring in their own moment but also do that before declaring the helpers package.

@splendido's suggestion on the forum to get your package into the official moment package roster would probably mitigate all these problems by automating releases.

**Comment by chicagogrooves** [comment]
IMHO, I don't think a package author should be expected to support "BYO" scenarios. You'd include a weak dependency on any new moment-clone package that comes out, to ensure you load after them? And people who load your package would have to follow the README to bring in a moment that you cannot run without ? 

I think you should set the expectation that you will declare the minimum version of moment you require, and add a hard dependency on that.

You can log some error messages if competing versions are detected, but your job is done once somebody loaded your lib and got it fully functioning with all your dependencies satisfied. 

Just don't let your lib become a bottleneck to getting newer versions of moment - thus the suggestion of depending on an official moment package sounds appealing.

**Comment by splendido** [comment]
My opinion is we should try to keep our *package directory* (i.e. Atmosphere as the only official interface) as clean as possible.
This means we should have **one** *official* package for each famous/valuable/interesting library.

A number of developer (me, @dandv @zimme and others included) have started a good initiative towards this direction: see [Official Meteor integration directly from 3rd party libraries](https://github.com/MeteorCommunity/discussions/issues/14), which was the first start, [MeteorPackaging](https://github.com/MeteorPackaging/discussions) as the place where put all *official* wrappers and last but not least [autopublish.meteor.com](http://autopublish.meteor.com/) which is the tool currently under development to keep all wrappers automatically up to date.

In this scenario I'd say everytime we have an *official* wrapper package, we should concentrate on it and leave others out working together as a community to solve problems and fix bugs (if any).

This said, I think `momentjs:moment` should be the only (hard) dependency for a package like `lbee:moment-helpers` even if `mrt:moment` is still being used by people (perhaps because the majority of us is not aware it's no more maintained by MDG).
What more, `mrt:moment` was left at version `2.8.1` which should be still compatible with latest version `2.10.3`: I see no problems for people using `mrt:moment` to switch to `momentjs:moment`.

In summary I'd probably depend on `momentjs:moment` and put a warning in case `mrt:moment` is detected rather than letting people chose what to use ;-)

**Comment by zimme** [comment]
When using weak dependencies the export is available like this `Package['momentjs:moment'].moment`.

So what I would suggest if you REALLY want to support several moment packages is to use weak dependencies and have a check that at least one of the weak dependencies are available and always use the "official" package if available.

If people have their own moment library they want to use they will just have to package it as a local package then just fork this package use it as a local package and add their local moment package as a weak dependency.

**Comment by lb-** [comment]
@zimme @splendido @chicagogrooves 

I think it would be best (& simplest) if this package simply relies on the official `momentjs:moment ` package via a standard dependency.

It will be much easier to maintain and for edge cases where people may feel they need to use a different momentjs package, that person can think about some work arounds or run their own local version of the package.

I feel that the trend in the Meteor ecosystem is definitely towards official packages and the variation is simply a legacy issue of the fast and recent growth in Meteor.

Regarding the package maintenance, I have put together an iftt that will create an issue on this repo whenever a new momentjs release is posted.

I will close this issue in a few days, unless people have some further thoughts.

Btw - thank you for your input, it is honestly great to have you guys post your thoughts, I have really appreciated your input in to the community (with posts, interviews, packages), thanks for making it fun.

**Comment by splendido** [comment]
out of curiosity, how is this "iftt" based/working?
...I'll have to do something similar for autopublish in a few days ;-)

**Comment by lb-** [comment]
@splendido here is the published recipe https://ifttt.com/recipes/292195-moment-js-releases-to-github-issue

however, it has not run yet (waiting on new feed item - next release), will see how it goes, but have had good experiences with ifttt previously.

**Comment by splendido** [comment]
I can see very little... perhaps I'd need to sign up ;-)
Thanks anyway, we might discuss about this in a near future...

**Comment by lb-** [comment]
@splendido ignore the ifttt - it is not working, I found a better way to do it with Zapier (free accounts lets you set up 10 zaps a month).

It generated this test issue: https://github.com/lb-/moment-helpers/issues/18

Basically
1. Sign up for Zapier
2. Connect your github account
3. Click 'make a zap' at the top
4. Trigger: Search for RSS & then 'new item in feed'
5. Action: Search for Github & then 'create an issue'
6. Click Continue a few times :smile: 
7. Paste in the url, the moment js one is this: https://github.com/moment/moment/releases.atom (essentially just put .atom at the end of a releases page), then continue again
8. The rest asks you about your issue options (you can put parts of the atom feed in the title & body of the issue, pretty nice. 
9. Continue & you can test it out, then finally turn it on.

**Comment by splendido** [comment]
Cool! :)


...but at this point, speaking about [autopublish.meteor.com](https://autopublish.meteor.com/), I see it much easier to poll a number of repositories for the latest release number once a week: see the [git hub API](https://developer.github.com/v3/repos/releases/#get-the-latest-release).
If you notice it's different from the latest published one, you have to trigger a new publish, that's it...

**Comment by lb-** [comment]
Sweet I will set that up at some point

**Comment by lb-** [comment]
Closing for now, will reopen if this requires further discussion.
